### PR TITLE
support compile-time for-loop unrolling

### PR DESCRIPTION
**Pull request by eternalphane**
This PR adds compile-time for-loop unrolling to CuPy JIT. The loop will be unrolled if both `start`, `stop`, `step` are `Constant`.

Example
```python
@jit.rawkernel()
def f(x, y):
    for i in range(5):
        y[i] = x[i]
x = cupy.ones(5, dtype=cupy.int32)
y = cupy.empty(5, dtype=cupy.int32)
f((1,), (1,), (x, y))
```
Generated kernel
```cpp
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  {
    y[0] = x[0];
  }
  {
    y[1] = x[1];
  }
  {
    y[2] = x[2];
  }
  {
    y[3] = x[3];
  }
  {
    y[4] = x[4];
  }
}
```
Don't know if it's necessary to add unit test...

### CODE REVIEW

**Comment by asi1024** [comment]
@eternalphane Thank you for your PR!
This PR always unrolls for-loops when the argument of `range` is a constant, but I am concerned that the huge size CUDA code will be generated when looping a huge number of times.

How about introducing `#pragma unroll`-like syntax in CuPy JIT likes as follows?

```py
@jit.rawkernel()
def f(x, y):
    with jit.unroll()
        for i in range(5):
            y[i] = x[i]
```

**Comment by asi1024** [comment]
There are two possible types of implementation to generate CUDA codes.

1. Unrolled in CuPy JIT (The one suggested)

```cpp
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  {
    y[0] = x[0];
  }
  {
    y[1] = x[1];
  }
   ...
```

2. Just use C++ `#pargma unroll`
```cpp
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  int i;
  #pragma unroll
  for (i = 0; i < 5; ++i) {
    y[i] = x[i];
  }
}
```

I think both ways are acceptable, but the first one is better because the variable `i` can be a compile-time constant.

**Comment by leofang** [comment]
`with jit.unroll(n)` is a nice idea. I'd prefer it to be 1:1 mapped to `#pragma unroll(n)` and let the C++ compiler make the decision for a given architecture (the compiler is free to ignore this instruction if it thinks it is not helping). It's easier to implement in CuPy JIT, and generate more readable code (which in turn is easier for debugging and reasoning). 

If we checkout out the Thrust/CUB repo, recently there are a few bug reports pointing out that aggressive unrolling (as is done in the current form e0d3f7c) actually hurts performance. Just my two cents.

**Comment by kmaehashi** [comment]
Thanks for working on this @eternalphane 😄 

* I guess NVCC automatically unrolls in cases like `range(5)`: https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll
* +1 to use `#pragma unroll` and let the compiler do the optimization.
* I'm unsure if `with` block fits here, as the unroll is effective only for the loop right after the pragma. Maybe 1:1 mapping from `jit.unroll()` -> `#pragma unroll` and `jit.unroll(n)` -> `#pragma unroll(n)` is better?

**Comment by leofang** [comment]
> * I'm unsure if `with` block fits here, as the unroll is effective only for the loop right after the pragma. Maybe 1:1 mapping from `jit.unroll()` -> `#pragma unroll` and `jit.unroll(n)` -> `#pragma unroll(n)` is better?

How about a language extension? 😂 We can add a keyword-only argument to `range`, so we
- map `for i in range(5, unroll=True)` to
```C
#pragma unroll
for (int i = 0; i < 5; i++) {
```
- map `for i in range(5, unroll=5)` to
```C
#pragma unroll(5)
for (int i = 0; i < 5; i++) {
```
- default to no `#pragma unroll` (which is the status quo)

I think this should be doable by extending the internal `Range`/`RangeFunc` objects.

**Comment by kmaehashi** [comment]
That's interesting, but I heard @asi1024 has an idea of implementing a CPU emulation in the future, so I guess it's better not to diverge from Python's language spec?

**Comment by asi1024** [comment]
CPU emulation also requires AST parse and cannot be performed by directly calling the target function. (For example, some threads may pass then-clause of the if statement, and other threads may pass else-clause.)

**Comment by asi1024** [comment]
Even if not so, I agree not to diverge from Python's language spec. How about supporting `cupyx.jit.range` for the extension?

**Comment by eternalphane** [comment]
I'm for supporting `cupyx.jit.range(unroll=n)`. Any other ideas? @asi1024 @kmaehashi  @leofang

**Comment by leofang** [comment]
As long as it is clearly documented (hopefully with an example) I think it's fine. We need a documentation to explain the loop unrolling feature anyway.

**Comment by eternalphane** [comment]
With `cupyx.jit.range`, the following code
```python
@jit.rawkernel()
def f(x, y):
    for i in jit.range(5):
        y[i] = x[i]
    for i in jit.range(5, unroll=True):
        y[i] = x[i]
    for i in jit.range(5, unroll=False):
        y[i] = x[i]
    for i in jit.range(5, unroll=5):
        y[i] = x[i]
x = cupy.ones(5, dtype=cupy.int32)
y = cupy.empty(5, dtype=cupy.int32)
f((1,), (1,), (x, y))
```
will be transpiled to
```cpp
extern "C" __global__ void f(CArray<int, 1, true, true> x, CArray<int, 1, true, true> y) {
  int i;
  for (int __it = 0, __stop = 5, __step = 1; __it < __stop; __it += __step) {
    i = __it;
    y[i] = x[i];
  }
  #pragma unroll
  for (int __it = 0, __stop = 5, __step = 1; __it < __stop; __it += __step) {
    i = __it;
    y[i] = x[i];
  }
  #pragma unroll(1)
  for (int __it = 0, __stop = 5, __step = 1; __it < __stop; __it += __step) {
    i = __it;
    y[i] = x[i];
  }
  #pragma unroll(5)
  for (int __it = 0, __stop = 5, __step = 1; __it < __stop; __it += __step) {
    i = __it;
    y[i] = x[i];
  }
}
```

File: cupyx/jit/_builtin_funcs.py [hunk 1]
```diff
@@ -15,6 +15,31 @@
...
+                If ``None``, leave the control of loop unrolling to compiler.
+
+        .. seealso:: `#pragma unroll`_
+
+        .. _#pragma unroll:
+            https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll
+        """
+        super.__call__(self)
+
     def call(self, env, *args, **kwargs):
```
Comment (leofang) [inline cupyx/jit/_builtin_funcs.py:43]: My impression is that @asi1024 prefers it to be explicit (https://github.com/cupy/cupy/pull/5335#discussion_r654654471):
```suggestion
    def call(self, env, *args, *, unroll=None):
```

File: cupyx/jit/_builtin_funcs.py [hunk 2]
```diff
@@ -15,6 +15,31 @@
...
+                If ``int`` type, add ``#pragma unroll(n)`` directive before the
+                loop, where ``n`` means the number of iterations to unroll.
+                If ``None``, leave the control of loop unrolling to compiler.
+
+        .. seealso:: `#pragma unroll`_
+
+        .. _#pragma unroll:
+            https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#pragma-unroll
+        """
+        super.__call__(self)
```
Comment (leofang) [inline cupyx/jit/_builtin_funcs.py:41]: Nitpick: technically this is incorrect (see https://github.com/cupy/cupy/pull/5361):
```suggestion
        super().__call__()
```

File: cupyx/jit/_builtin_funcs.py [hunk 3]
```diff
@@ -28,6 +53,25 @@ def call(self, env, *args, **kwargs):
             raise TypeError(
                 f'range expected at most 3 argument, got {len(args)}')
 
+        unroll = kwargs.get('unroll')
```
Comment (leofang) [inline cupyx/jit/_builtin_funcs.py:56]: Together with the change above, this can be removed:
```suggestion
```

File: cupyx/jit/_builtin_funcs.py [hunk 4]
```diff
@@ -28,6 +53,25 @@ def call(self, env, *args, **kwargs):
...
+                raise TypeError(
+                    'loop unrolling requires constant start, stop, step and '
+                    'unroll value')
+            unroll = unroll.obj
+            if not (isinstance(unroll, int) or isinstance(unroll, bool)):
+                raise TypeError(
+                    'unroll value expected to be of type int, '
+                    f'got {type(unroll).__name__}')
+            if unroll is False:
+                unroll = 1
```
Comment (leofang) [inline cupyx/jit/_builtin_funcs.py:69]: I am not sure how useful it is to set `#pragma unroll(1)`, at least I haven't encountered such need in real life yet. Let's have @asi1024 make the final call 🙂

**Review by leofang — CHANGES_REQUESTED** [review]

File: cupyx/jit/_builtin_funcs.py (hunk 4 above)
Comment (eternalphane) [reply cupyx/jit/_builtin_funcs.py:68]: @kmaehashi [mentions in the first bullet](https://github.com/cupy/cupy/pull/5343#issuecomment-862842340) that the CUDA doc states:
> By default, the compiler unrolls small loops with a known trip count.

I agree it's not a common use case, but `#pragma unroll(1)` explicitly disables unrolling, which is necessary for users to take full control of the unrolling behavior.

File: cupyx/jit/_builtin_funcs.py (hunk 1 above)
Comment (eternalphane) [reply cupyx/jit/_builtin_funcs.py]: `unroll` is already a keyword-only argument, note the `*args` before `unroll` 😄

File: cupyx/jit/_builtin_funcs.py (hunk 4 above)
Comment (asi1024) [reply cupyx/jit/_builtin_funcs.py:68]: The users can control it by specifying `unroll=1` instead.

Comment (kmaehashi) [reply cupyx/jit/_builtin_funcs.py:68]: How about mapping `False` to `#pragma nounroll`?
https://github.com/NVlabs/xmp/blob/0052dbb206ef0cb2b47b0fe5d61f54e5cd2fa73c/src/include/digitmp/mul.h#L42

Comment (asi1024) [reply cupyx/jit/_builtin_funcs.py:68]: I prefer using `"always"`, `"default"`, `"never"` literals instead of `True`, `None` and `False`.

Comment (kmaehashi) [reply cupyx/jit/_builtin_funcs.py:68]: @asi1024 As `unroll` is intended for users understanding `#pragma unroll` behavior, I think these mappings are enough clear and straightforward, so +1 not to introduce our own semantics on the argument.
* `unroll=True`: `#pragma unroll`
* `unroll=False`:  `#pragma nounroll`
* `unroll=3`:  `#pragma unroll(3)`
* not specified (or `unroll=None`): `` (no pragma)
* others: error

**Comment by asi1024** [comment]
Could you add tests?

File: cupyx/jit/_internal_types.py [hunk 5]
```diff
@@ -47,12 +47,13 @@ def __repr__(self):
 
 class Range(Expr):
 
-    def __init__(self, start, stop, step, ctype, step_is_positive):
+    def __init__(self, start, stop, step, ctype, step_is_positive, unroll):
```
Comment (asi1024) [inline cupyx/jit/_internal_types.py:50]: This fix looks to change the behavior also when using the built-in range.

**Comment by leofang** [comment]
Hi @eternalphane would you have time to finish this PR, or should one of us take over?

File: cupyx/jit/_builtin_funcs.py (hunk 4 above)
Comment (leofang) [reply cupyx/jit/_builtin_funcs.py:68]: Sorry, I was being silly. @eternalphane was right that `#pragma unroll(1)` is the correct directive to disable unrolling. This is documented in the Programming Guide. 

`#pragma nounroll` has no effect (`nounroll` should be done at the PTX level, not the C++ level). 

We can do a little experiment with https://godbolt.org/z/soanE1hhv by observing the output PTX length for different pragma setups.

**Pull request closed without merging**